scipy/_lib/doccer.py

```diff
@@ -1,26 +1,40 @@
-''' Utilities to allow inserting docstring fragments for common
-parameters into function and method docstrings'''
+"""Utilities to allow inserting docstring fragments for common
+parameters into function and method docstrings."""
 
+from collections.abc import Callable, Iterable, Mapping
+from typing import Protocol, TypeVar
 import sys
 
 __all__ = [
-    'docformat', 'inherit_docstring_from', 'indentcount_lines',
-    'filldoc', 'unindent_dict', 'unindent_string', 'extend_notes_in_docstring',
-    'replace_notes_in_docstring', 'doc_replace'
+    "docformat",
+    "inherit_docstring_from",
+    "indentcount_lines",
+    "filldoc",
+    "unindent_dict",
+    "unindent_string",
+    "extend_notes_in_docstring",
+    "replace_notes_in_docstring",
+    "doc_replace",
 ]
 
+_F = TypeVar("_F", bound=Callable[..., object])
 
-def docformat(docstring, docdict=None):
-    ''' Fill a function docstring from variables in dictionary
+
+class Decorator(Protocol):
```

Comment:
Will this be part of the public api? If so, then it probably needs a docstring. Otherwise, it should probably be prefixed with a `_`.

Reply:
This file is under `_lib` so will not be a part of the public API. I think it doesn't really matter either way whether a leading underscore is used here or not ([IMO PEP8 is a bit ambiguous](https://github.com/astral-sh/ruff/issues/10755)). A one-line (or more, if there is more to say) docstring is always nice, though :)

Reply:
Ah of course, we're already operating within the covers of the `_` prefix; my bad 🙃

Reply:
Thanks for reminding me about docstrings! I added a simple one because I'm not entirely sure how to properly document such simple protocols. Tell me if you want me to be more informative and less self-referential (I used the word decorator to describe the decorator protocol haha).

I feel like decorators should just be part of the typing module instead of having to create a custom protocol.

Reply:
Maybe it'll help to point out that this protocol is (only) meant for static typing purposes. Then I suppose it should be clear enough 👌🏻.

> I feel like decorators should just be part of the typing module instead of having to create a custom protocol.

Yea I feel you you... Unfortunately, "DRY" is a virtually impossible to achieve within python typing in its current form